File: python-flask-api/app/controllers/assignments.py
from flask import Blueprint
from sqlalchemy import text
from app import db, config, jsonify, request
from ..helpers.http_errors import InternalServerError, BadRequest, ResourceNotFound
from ..helpers.utils import utils
from ..models.assignments import *


Assignments_blueprint = Blueprint('Assignments', __name__)


# list Assignments records
# @param fieldname: filter table records by a field
# @param fieldvalue:  filter value
# @request.args search: search records
# @request.args orderby: sort records by field name
# @request.args ordertype: sort type (asc|desc)
# @request.args page: current page number
# @request.args limit: maximum number of records to select
@Assignments_blueprint.route('/')
@Assignments_blueprint.route('/index/<fieldname>/<fieldvalue>')
def list(fieldname=None, fieldvalue=None):
    try:
        query = Assignments.query
        search = request.args.get('search')
        if search:
            query = query.filter(
                Assignments.year.like(f'%{search}%') | 
                Assignments.term.like(f'%{search}%') | 
                Assignments.subject.like(f'%{search}%') | 
                Assignments.asignment_name.like(f'%{search}%') | 
                Assignments.id.like(f'%{search}%') | 
                Assignments.instructions.like(f'%{search}%') 
            )
        
        # filter by dynamic field name
        if fieldname:
            field_filter = text(f'{fieldname} = :fieldvalue').params(fieldvalue=fieldvalue)
            query = query.filter(field_filter)
        
        orderby = request.args.get('orderby')
        ordertype = request.args.get('ordertype', 'desc')
        if orderby:
            query = query.order_by(text(f'{orderby} {ordertype}'))
        else:
            query = query.order_by(text(f'assignments.id {ordertype}'))
        
        # fields to select
        query = query.with_entities(
            Assignments.year,
            Assignments.term,
            Assignments.subject,
            Assignments.asignment_name,
            Assignments.id
        )
        page = int(request.args.get('page', 1))
        limit = int(request.args.get('limit', config.MAX_RECORD_COUNT))
        offset = ((page-1) * limit)
        total_records = query.count()
        records = query.limit(limit).offset(offset).all()
        record_count = len(records)
        total_pages = round(total_records/limit)
        records = [(row._asdict()) for row in records]
        
        # response object
        response = dict(
            records=records, 
            totalRecords=total_records,
            recordCount=record_count,
            totalPages=total_pages
        )
        return jsonify(response)
    except Exception as ex:
        return InternalServerError(ex)


# Select table record by ID
@Assignments_blueprint.route('/view/<rec_id>')
def view(rec_id=None):
    try:
        query = Assignments.query
        query = query.filter(Assignments.id == rec_id)
        query = query.with_entities(
            Assignments.id,
            Assignments.year,
            Assignments.term,
            Assignments.subject,
            Assignments.asignment_name,
            Assignments.file,
            Assignments.instructions
        )
        
        record = query.first()
        if not record: return ResourceNotFound()
         
        record = record._asdict()
        
        # return result as json
        return jsonify(record)
        
    except Exception as ex:
        return InternalServerError(ex)


# Save form data to the Assignments table
@Assignments_blueprint.route('/add', methods=['POST'])
def add():
    try:
        modeldata = request.body
        
        # move uploaded file from temp directory to destination directory
        if "file" in modeldata:
            file_info = utils.move_uploaded_files(modeldata['file'], "file")
            modeldata['file'] = file_info['filepath']
         
        form = AssignmentsAddForm(modeldata)
        errors = [] # list of validation errors
        
        # validate assignments form data
        if not form.validate():
            errors.append(form.errors)
        
        if errors:
            return BadRequest(errors)
        
        record = Assignments()
        form.populate_obj(record)
        
        # save assignments records
        db.session.add(record)
        db.session.commit()
        db.session.flush()
        rec_id = record.id
         
        record = record._asdict()
        return jsonify(record)
    except Exception as ex:
        return InternalServerError(ex)


# Select record by table primary key and update with form data
@Assignments_blueprint.route('/edit/<rec_id>', methods=['GET', 'POST'])
def edit(rec_id=None):
    try:
        query = Assignments.query
        query = query.filter(Assignments.id == rec_id)
        record = query.first()
        if not record: return ResourceNotFound()
        
        if request.method == 'POST':
            errors = []
            modeldata = request.body
            
            # move uploaded file from temp directory to destination directory
            if "file" in modeldata:
                file_info = utils.move_uploaded_files(modeldata['file'], "file")
                modeldata['file'] = file_info['filepath']
            
            form = AssignmentsEditForm(modeldata, obj=record)
            
            if not form.validate():
                errors.append(form.errors)
            
            if errors:
                return BadRequest(errors)
            
            # save Assignments record
            form.populate_obj(record)
            db.session.commit()
         
        record = record._asdict()
        return jsonify(record)
    except Exception as ex:
        return InternalServerError(ex)


# Delete record from the database
# Support multi delete by separating record id by comma.
@Assignments_blueprint.route('/delete/<rec_id>')
def delete(rec_id):
    query = Assignments.query
    arr_id = rec_id.split(',')
    try:
        query = query.filter(Assignments.id.in_(arr_id))
        query.delete(synchronize_session=False)
        db.session.commit()
        
        return jsonify(arr_id)
    except Exception as ex:
        return InternalServerError(ex)

File: nodejs-express-api/controllers/assignments.js
import { Router } from 'express';
import { body } from 'express-validator';
import uploader from '../helpers/uploader.js';
import validateFormData from '../helpers/validate_form.js';
import DB from '../models/db.js';


const router = Router();




/**
 * Route to list assignments records
 * @GET /assignments/index/{fieldname}/{fieldvalue}
 */
router.get(['/', '/index/:fieldname?/:fieldvalue?'], async (req, res) => {  
	try{
		const query = {};
		let queryFilters = [];
		let where = {};
		let replacements = {};
		let fieldName = req.params.fieldname;
		let fieldValue = req.params.fieldvalue;
		
		if (fieldName){
			queryFilters.push(DB.filterBy(fieldName, fieldValue));
		}
		let search = req.query.search;
		if(search){
			let searchFields = DB.Assignments.searchFields();
			where[DB.op.or] = searchFields;
			replacements.search = `%${search}%`;
		}
		
		if(queryFilters.length){
			where[DB.op.and] = queryFilters;
		}
		query.raw = true;
		query.where = where;
		query.replacements = replacements;
		query.order = DB.getOrderBy(req, 'id', 'desc');
		query.attributes = DB.Assignments.listFields();
		let page = parseInt(req.query.page) || 1;
		let limit = parseInt(req.query.limit) || 10;
		let result = await DB.Assignments.paginate(query, page, limit);
		return res.ok(result);
	}
	catch(err) {
		return res.serverError(err);
	}
});


/**
 * Route to view Assignments record
 * @GET /assignments/view/{recid}
 */
router.get('/view/:recid', async (req, res) => {
	try{
		const recid = req.params.recid || null;
		const query = {}
		const where = {}
		where['id'] = recid;
		query.raw = true;
		query.where = where;
		query.attributes = DB.Assignments.viewFields();
		let record = await DB.Assignments.findOne(query);
		if(!record){
			return res.notFound();
		}
		return res.ok(record);
	}
	catch(err){
		return res.serverError(err);
	}
});


/**
 * Route to insert Assignments record
 * @POST /assignments/add
 */
router.post('/add/', 
	[
		body('year').optional({nullable: true, checkFalsy: true}),
		body('term').not().isEmpty(),
		body('subject').not().isEmpty(),
		body('asignment_name').not().isEmpty(),
		body('file').not().isEmpty(),
		body('instructions').not().isEmpty(),
	], validateFormData
, async function (req, res) {
	try{
		let modeldata = req.getValidFormData();
		
        // move uploaded file from temp directory to destination directory
		if(modeldata.file !== undefined) {
			const fileInfo = uploader.moveUploadedFiles(modeldata.file, 'file');
			modeldata.file = fileInfo.filepath;
		}
		
		//save Assignments record
		let record = await DB.Assignments.create(modeldata);
		//await record.reload(); //reload the record from database
		const recid =  record['id'];
		const newValues = JSON.stringify(record); 
		req.writeToAuditLog({ recid, oldValues: null, newValues });
		
		return res.ok(record);
	} catch(err){
		return res.serverError(err);
	}
});


/**
 * Route to get  Assignments record for edit
 * @GET /assignments/edit/{recid}
 */
router.get('/edit/:recid', async (req, res) => {
	try{
		const recid = req.params.recid;
		const query = {};
		const where = {};
		where['id'] = recid;
		query.raw = true;
		query.where = where;
		query.attributes = DB.Assignments.editFields();
		let record = await DB.Assignments.findOne(query);
		if(!record){
			return res.notFound();
		}
		return res.ok(record);
	}
	catch(err){
		return res.serverError(err);
	}
});


/**
 * Route to update  Assignments record
 * @POST /assignments/edit/{recid}
 */
router.post('/edit/:recid', 
	[
		body('year').optional({nullable: true, checkFalsy: true}),
		body('term').optional({nullable: true}).not().isEmpty(),
		body('subject').optional({nullable: true}).not().isEmpty(),
		body('asignment_name').optional({nullable: true}).not().isEmpty(),
		body('file').optional({nullable: true}).not().isEmpty(),
		body('instructions').optional({nullable: true}).not().isEmpty(),
	], validateFormData
, async (req, res) => {
	try{
		const recid = req.params.recid;
		let modeldata = req.getValidFormData({ includeOptionals: true });
		
        // move uploaded file from temp directory to destination directory
		if(modeldata.file !== undefined) {
			const fileInfo = uploader.moveUploadedFiles(modeldata.file, 'file');
			modeldata.file = fileInfo.filepath;
		}
		const query = {};
		const where = {};
		where['id'] = recid;
		query.raw = true;
		query.where = where;
		query.attributes = DB.Assignments.editFields();
		let record = await DB.Assignments.findOne(query);
		if(!record){
			return res.notFound();
		}
		const oldValues = JSON.stringify(record); //for audit trail
		await DB.Assignments.update(modeldata, {where: where});
		record = await DB.Assignments.findOne(query);//for audit trail
		const newValues = JSON.stringify(record); 
		req.writeToAuditLog({ recid, oldValues, newValues });
		return res.ok(modeldata);
	}
	catch(err){
		return res.serverError(err);
	}
});


/**
 * Route to delete Assignments record by table primary key
 * Multi delete supported by recid separated by comma(,)
 * @GET /assignments/delete/{recid}
 */
router.get('/delete/:recid', async (req, res) => {
	try{
		const recid = (req.params.recid || '').split(',');
		const query = {};
		const where = {};
		where['id'] = recid;
		query.raw = true;
		query.where = where;
		let records = await DB.Assignments.findAll(query);
		records.forEach(async (record) => { 
			//perform action on each record before delete
			const oldValues = JSON.stringify(record); //for audit trail
			req.writeToAuditLog({ recid: record['id'], oldValues });
		});
		await DB.Assignments.destroy(query);
		return res.ok(recid);
	}
	catch(err){
		return res.serverError(err);
	}
});
export default router;
